File: src/main/java/com/jotoai/voenix/shop/prompts/entity/Prompt.java
package com.jotoai.voenix.shop.prompts.entity;

import jakarta.persistence.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.Objects;

@Entity
@Table(name = "prompts")
public class Prompt {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(nullable = false, length = 500)
    private String title;
    
    @Column(columnDefinition = "TEXT")
    private String content;
    
    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @UpdateTimestamp
    @Column(nullable = false)
    private LocalDateTime updatedAt;

    // Constructors
    public Prompt() {
    }

    public Prompt(Long id, String title, String content, LocalDateTime createdAt, LocalDateTime updatedAt) {
        this.id = id;
        this.title = title;
        this.content = content;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }

    // Builder pattern
    public static PromptBuilder builder() {
        return new PromptBuilder();
    }

    // Getters and Setters
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    // equals and hashCode
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Prompt prompt = (Prompt) o;
        return Objects.equals(id, prompt.id) &&
                Objects.equals(title, prompt.title) &&
                Objects.equals(content, prompt.content) &&
                Objects.equals(createdAt, prompt.createdAt) &&
                Objects.equals(updatedAt, prompt.updatedAt);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, title, content, createdAt, updatedAt);
    }

    // toString
    @Override
    public String toString() {
        return "Prompt{" +
                "id=" + id +
                ", title='" + title + '\'' +
                ", content='" + content + '\'' +
                ", createdAt=" + createdAt +
                ", updatedAt=" + updatedAt +
                '}';
    }

    // Builder class
    public static class PromptBuilder {
        private Long id;
        private String title;
        private String content;
        private LocalDateTime createdAt;
        private LocalDateTime updatedAt;

        PromptBuilder() {
        }

        public PromptBuilder id(Long id) {
            this.id = id;
            return this;
        }

        public PromptBuilder title(String title) {
            this.title = title;
            return this;
        }

        public PromptBuilder content(String content) {
            this.content = content;
            return this;
        }

        public PromptBuilder createdAt(LocalDateTime createdAt) {
            this.createdAt = createdAt;
            return this;
        }

        public PromptBuilder updatedAt(LocalDateTime updatedAt) {
            this.updatedAt = updatedAt;
            return this;
        }

        public Prompt build() {
            return new Prompt(id, title, content, createdAt, updatedAt);
        }
    }
}

File: backend/src/main/kotlin/com/jotoai/voenix/shop/domain/prompts/entity/Prompt.kt
package com.jotoai.voenix.shop.domain.prompts.entity

import com.jotoai.voenix.shop.domain.prompts.dto.PromptDto
import com.jotoai.voenix.shop.domain.prompts.dto.PublicPromptDto
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.OffsetDateTime

@Entity
@Table(name = "prompts")
data class Prompt(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,
    @Column(nullable = false, length = 500)
    var title: String,
    @Column(name = "prompt_text", columnDefinition = "TEXT")
    var promptText: String? = null,
    @Column(name = "category_id")
    var categoryId: Long? = null,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", insertable = false, updatable = false)
    var category: PromptCategory? = null,
    @Column(name = "subcategory_id")
    var subcategoryId: Long? = null,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "subcategory_id", insertable = false, updatable = false)
    var subcategory: PromptSubCategory? = null,
    @Column(nullable = false)
    var active: Boolean = true,
    @Column(name = "example_image_filename", length = 500)
    var exampleImageFilename: String? = null,
    @OneToMany(mappedBy = "prompt", cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.LAZY)
    var promptSlotVariantMappings: MutableList<PromptSlotVariantMapping> = mutableListOf(),
    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false, columnDefinition = "timestamptz")
    val createdAt: OffsetDateTime? = null,
    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false, columnDefinition = "timestamptz")
    var updatedAt: OffsetDateTime? = null,
) {
    fun toDto() =
        PromptDto(
            id = requireNotNull(this.id) { "Prompt ID cannot be null when converting to DTO" },
            title = this.title,
            promptText = this.promptText,
            categoryId = this.categoryId,
            category = this.category?.toDto(),
            subcategoryId = this.subcategoryId,
            subcategory = this.subcategory?.toDto(),
            active = this.active,
            slots =
                this.promptSlotVariantMappings
                    .sortedBy {
                        it.promptSlotVariant.promptSlotType?.position ?: 0
                    }.map {
                        it.promptSlotVariant
                            .toDto()
                    },
            exampleImageUrl = this.exampleImageFilename?.let { "/images/prompt-example-images/$it" },
            createdAt = this.createdAt,
            updatedAt = this.updatedAt,
        )

    fun addPromptSlotVariant(promptSlotVariant: PromptSlotVariant) {
        val mapping =
            PromptSlotVariantMapping(
                id = PromptSlotVariantMappingId(this.id ?: 0, promptSlotVariant.id ?: 0),
                prompt = this,
                promptSlotVariant = promptSlotVariant,
            )
        this.promptSlotVariantMappings.add(mapping)
    }

    fun removePromptSlotVariant(promptSlotVariant: PromptSlotVariant) {
        this.promptSlotVariantMappings.removeIf { it.promptSlotVariant.id == promptSlotVariant.id }
    }

    fun clearPromptSlotVariants() {
        this.promptSlotVariantMappings.clear()
    }

    fun toPublicDto() =
        PublicPromptDto(
            id = requireNotNull(this.id) { "Prompt ID cannot be null when converting to DTO" },
            title = this.title,
            exampleImageUrl = this.exampleImageFilename?.let { "/images/prompt-example-images/$it" },
            category = this.category?.toPublicDto(),
            subcategory = this.subcategory?.toPublicDto(),
            slots =
                this.promptSlotVariantMappings
                    .sortedBy {
                        it.promptSlotVariant.promptSlotType?.position ?: 0
                    }.map {
                        it.promptSlotVariant
                            .toPublicDto()
                    },
        )
}
